src/v/cluster/controller.cc

```diff
@@ -246,11 +246,48 @@ ss::future<> controller::start() {
           return _stm.invoke_on(controller_stm_shard, &controller_stm::start);
       })
       .then([this] {
-          return _stm.invoke_on(controller_stm_shard, [](controller_stm& stm) {
-              // we do not have to use timeout in here as all the batches to
-              // apply have to be accesssible
-              return stm.wait(stm.bootstrap_last_applied(), model::no_timeout);
-          });
+          auto disk_commit_offset = _raft0->log().offsets().committed_offset;
+
+          return _stm
+            .invoke_on(
+              controller_stm_shard,
+              [disk_commit_offset](controller_stm& stm) {
+                  // we do not have to use timeout in here as all the batches to
+                  // apply have to be accesssible
+                  auto last_applied = stm.bootstrap_last_applied();
+
+                  // Consistency check: on a bug-free system, the last_applied
+                  // value in the kvstore always points to data on disk.
+                  // However, if there is a bug, or someone has removed some log
+                  // segments out of band, we will hang trying to read up to
+                  // last_applied. Mitigate this by clamping it to the top of
+                  // the log on disk.
+                  if (last_applied > disk_commit_offset) {
+                      vlog(
+                        clusterlog.error,
+                        "Inconsistency detected between KVStore last_applied "
+                        "({}) and actual log size ({}).  If the storage "
+                        "directory was not modified intentionally, this is a "
+                        "bug.",
+                        last_applied,
+                        disk_commit_offset);
+                      last_applied = std::max(
```

Comment:
we should not always wait for `disk_commit_offset` it may be the case that some of the entries will not be applied at all

Reply:
I've modified this to use a timeout: that way we would usually do a nice orderly startup by waiting for disk_commit_offset, but in the event that there was a truncation the timeout will prevent us from hanging.